GoogleSignIn/Tests/Unit/GIDVerifyAccountDetailTest.m

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <XCTest/XCTest.h>
+
+#if TARGET_OS_IOS
+#import "GoogleSignIn/Sources/Public/GoogleSignIn/GIDConfiguration.h"
+#import "GoogleSignIn/Sources/Public/GoogleSignIn/GIDVerifyAccountDetail.h"
+#import "GoogleSignIn/Sources/Public/GoogleSignIn/GIDVerifiableAccountDetail.h"
+
+#import "GoogleSignIn/Tests/Unit/GIDFakeMainBundle.h"
+
+static NSString * const kClientId = @"FakeClientID";
+static NSString * const kServerClientId = @"FakeServerClientID";
+static NSString * const kOpenIDRealm = @"FakeRealm";
+static NSString * const kFakeHostedDomain = @"fakehosteddomain.com";
+
+@interface GIDVerifyAccountDetailTests : XCTestCase
+// The |UIViewController| object being tested.
+@property UIViewController *presentingViewController;
+
+// Fake [NSBundle mainBundle].
+@property GIDFakeMainBundle *fakeMainBundle;
+
+// The |GIDVerifyAccountDetail| object being tested.
+@property GIDVerifyAccountDetail *verifyAccountDetail;
+
+// The list of account details when testing [GIDVerifiableAccountDetail].
+@property NSArray<GIDVerifiableAccountDetail *> *verifiableAccountDetails;
+@end
+
+@implementation GIDVerifyAccountDetailTests
+
+#pragma mark - Lifecycle
+
+- (void)setUp {
+  [super setUp];
+
+  _presentingViewController = [[UIViewController alloc] init];
+
+  _verifyAccountDetail = [[GIDVerifyAccountDetail alloc] init];
+
+  GIDVerifiableAccountDetail *ageOver18Detail = [[GIDVerifiableAccountDetail alloc] initWithAccountDetailType:GIDAccountDetailTypeAgeOver18];
+  _verifiableAccountDetails = @[ageOver18Detail];
+
+  _fakeMainBundle = [[GIDFakeMainBundle alloc] init];
+}
+
+
+#pragma mark - Tests
+
+- (void)testInit {
+  [_fakeMainBundle startFakingWithClientID:kClientId];
+
+  GIDVerifyAccountDetail *verifyAccountDetail = [[GIDVerifyAccountDetail alloc] init];
+  XCTAssertNotNil(verifyAccountDetail.configuration);
+  XCTAssertEqual(verifyAccountDetail.configuration.clientID, kClientId);
+  XCTAssertNil(verifyAccountDetail.configuration.serverClientID);
+  XCTAssertNil(verifyAccountDetail.configuration.hostedDomain);
+  XCTAssertNil(verifyAccountDetail.configuration.openIDRealm);
+}
+
+- (void)testInit_noConfig {
+  [_fakeMainBundle startFakingWithClientID:kClientId];
```

Comment:
I think you could pass nil for the client ID here and skip what's below. `GIDFakeMainBundle.h` doesn't have nullability annotations...

Reply:
Unable to pass nil for the client ID here since an exception will be thrown in [GIDFakeMainBundle](https://github.com/google/GoogleSignIn-iOS/blob/ec20caf39917e3a50905776db1713b37c6b33840/GoogleSignIn/Tests/Unit/GIDFakeMainBundle.m#L49) and . This is because `NSDictionary` does not allow nil values so it throws an exception on initialization. I believe that's why `kClientId` should be passed in to first initialize and then the values in the dictionary can be set to nil.

Reply:
An alternative could be to remove line `[_fakeMainBundle startFakingWithClientID:kClientId];` and only keep the second call but I understand this still runs into the issue with nullability annotations. Let me know what you think.

Reply:
You could update `GIDFakeMainBundle` to take a `nullable` string for the `clientId`:

```
- (void)startFakingWithClientID:(nullable NSString *)clientId {
  _clientId = clientId;

  _fakedKeys = @[ kCFBundleURLTypesKey,
                  kConfigClientIDKey,
                  kConfigServerClientIDKey,
                  kConfigHostedDomainKey,
                  kConfigOpenIDRealmKey ];

  _fakeConfig = clientId ? [@{ @"GIDClientID": clientId } mutableCopy] : [@{} mutableCopy];

  [GULSwizzler swizzleClass:[NSBundle class]
                   selector:@selector(objectForInfoDictionaryKey:)
            isClassSelector:NO
                  withBlock:^id(id _self, NSString *key) {
    if ([self->_fakedKeys containsObject:key]) {
      return self->_fakeConfig[key];
    } else {
      @throw [NSException exceptionWithName:@"Requested unexpected info.plist key."
                                     reason:nil
                                   userInfo:nil];
    }
  }];
}
```

Reply:
Good idea, done!

Reply:
Updated!